app/src/main/java/org/mozilla/fenix/home/recentvisits/RecentVisitsFeature.kt

```diff
@@ -0,0 +1,258 @@
+                    }
+                }
+            }
+            .map {
+                FeatureHistoryGroup(
+                    groupName = it.key,
+                    groupItems = it.value
+                )
+            }
+    }
+
+    @VisibleForTesting
+    internal fun updateState(
+        historyHighlights: List<FeatureHistoryHighlight>,
+        historyGroups: List<FeatureHistoryGroup>
+    ) {
+        homeStore.dispatch(
+            HomeFragmentAction.RecentHistoryChange(
+                getCombinedHistory(historyHighlights, historyGroups)
+            )
+        )
+    }
+
+    /**
+     * Get up to [MAX_RESULTS_TOTAL] items if available as an even split of history highlights and history groups.
+     * If more items then needed are available then highlights will be more by one.
+     *
+     * @param historyHighlights Sorted list of history highlights. Can be empty.
+     * @param historyGroups Sorted list of history groups. Can be empty.
+     */
+    @VisibleForTesting
+    internal fun getCombinedHistory(
+        historyHighlights: List<FeatureHistoryHighlight>,
+        historyGroups: List<FeatureHistoryGroup>
+    ): List<RecentlyVisitedItem> {
+        val totalItemsCount = historyHighlights.size + historyGroups.size
+
+        return if (totalItemsCount <= MAX_RESULTS_TOTAL) {
+            getSortedHistory(
+                historyHighlights.filterByMetadata(historyGroups),
+                historyGroups
+            )
+        } else {
+            var groupsCount = 0
+            var highlightCount = 0
+            while ((highlightCount + groupsCount) < MAX_RESULTS_TOTAL) {
+                if ((highlightCount + groupsCount) < MAX_RESULTS_TOTAL &&
+                    historyHighlights.getOrNull(highlightCount) != null
+                ) {
+                    highlightCount += 1
+                }
+
+                if ((highlightCount + groupsCount) < MAX_RESULTS_TOTAL &&
+                    historyGroups.getOrNull(groupsCount) != null
+                ) {
+                    groupsCount += 1
+                }
+            }
+
+            getSortedHistory(
+                historyHighlights
+                    .filterByMetadata(historyGroups)
+                    .sortedByDescending { it.lastAccessedTime }
+                    .take(highlightCount),
+                historyGroups
+                    .sortedByDescending { it.lastAccessedTime }
+                    .take(groupsCount)
+            )
+        }
+    }
+
+    @VisibleForTesting
+    internal fun getSortedHistory(
+        historyHighlights: List<FeatureHistoryHighlight>,
+        historyGroups: List<FeatureHistoryGroup>
+    ): List<RecentlyVisitedItem> {
+        return (historyHighlights + historyGroups)
+            .sortedByDescending { it.lastAccessedTime }
+            .map { when(it) {
+                is FeatureHistoryHighlight -> RecentHistoryHighlight(
+                    title = it.historyHighlight.title ?: it.historyHighlight.url,
```

Comment:
On the title here we should check for nullOrEmpty. Did some manual testing and had some empty entries for title. So, let's fall back to the URL with the title is an empty String.

Reply:
Previously this was the case and had a test for it.
Added now also.